lib/intercept.js

```diff
@@ -365,75 +368,105 @@ function activeMocks() {
+    headers: Object.fromEntries(fetchRequest.headers.entries())
+  };
+
+  const clientRequest = new http.ClientRequest(options);
+  // Note: You won't have access to the request body data from the Fetch Request
+
+  return clientRequest;
+}
 
-    const interceptors = interceptorsFor(options)
+function activate() {
+  if (originalClientRequest) {
+    throw new Error('Nock already active')
+  }
 
-    if (isOn() && interceptors) {
-      const matches = interceptors.some(interceptor =>
-        interceptor.matchOrigin(options)
-      )
-      const allowUnmocked = interceptors.some(
-        interceptor => interceptor.options.allowUnmocked
-      )
+  overrideClientRequest()
+  const interceptor = new BatchInterceptor({
+    name: 'my-interceptor',
+    interceptors: [...nodeInterceptors, new FetchInterceptor()],
+  })
+  interceptor.apply();
+  interceptor.on('request', function ({ request, requestId }) {
+    return new Promise((resolve) => {
+      const { options, callback } = common.normalizeClientRequestArgs(request.url)
+      options.proto = options.protocol.slice(0, -1)
+      const interceptors = interceptorsFor(options)
+      if (isOn() && interceptors) {
+        const matches = interceptors.some(interceptor =>
+          interceptor.matchOrigin(options)
+        )
+        const allowUnmocked = interceptors.some(
+          interceptor => interceptor.options.allowUnmocked
+        )
 
-      if (!matches && allowUnmocked) {
-        let req
-        if (proto === 'https') {
-          const { ClientRequest } = http
-          http.ClientRequest = originalClientRequest
-          req = overriddenRequest(options, callback)
-          http.ClientRequest = ClientRequest
-        } else {
-          req = overriddenRequest(options, callback)
+        if (!matches && allowUnmocked) {
+          // TODO: implement unmocked
+          // let req
+          // if (proto === 'https') {
+          //   const { ClientRequest } = http
+          //   http.ClientRequest = originalClientRequest
+          //   req = overriddenRequest(options, callback)
+          //   http.ClientRequest = ClientRequest
+          // } else {
+          //   req = overriddenRequest(options, callback)
+          // }
+          // globalEmitter.emit('no match', req)
+          // return req
+          throw new Error('TODO')
         }
-        globalEmitter.emit('no match', req)
-        return req
-      }
 
-      //  NOTE: Since we already overrode the http.ClientRequest we are in fact constructing
-      //    our own OverriddenClientRequest.
-      return new http.ClientRequest(options, callback)
-    } else {
-      globalEmitter.emit('no match', options)
-      if (isOff() || isEnabledForNetConnect(options)) {
-        return overriddenRequest(options, callback)
+        const req = convertFetchRequestToClientRequest(request);
+        req.on('response', response => {
+          request.respondWith(new Response('test', { status: 200 }))
```

Comment:
Nitpick: we'd want to read the response stream to the `ReadableStream` of the `response` instance and call `request.respondWith()` on the `end` event of the original response here. 

As I mentioned in the issue, let me know if exporting this existing utility from Interceptors would help here. I think it would.

Reply:
Do you have an idea how we can wait for the nockResponse `end` event? we can use promise, but maybe you had something else in mind.

Reply:
In this case, we can rely on the internal `http.ClientRequest` implementation to guarantee the order of event execution. 

```js
const stream = new Readable()
const fetchResponse = new Response(stream, {
  status: response.statusCode,
  statusText: response.statusMessage,
  headers: response.headers // pseudo-code
})

response.on('data', (chunk) => stream.write(chunk))
response.on('end', () => {
  stream.finish()
  request.respndWith(fetchResponse)
})
```

This is a gist. If you want the actual implementation example, take a look at [this function](https://github.com/mswjs/interceptors/blob/04152ed914f8041272b6e92ed374216b8177e1b2/src/interceptors/ClientRequest/utils/createResponse.ts#L7).